rubix-spi/src/main/java/com/qubole/rubix/spi/fop/ObjectPoolPartition.java

```diff
@@ -50,115 +53,83 @@ public ObjectPoolPartition(ObjectPool<T> pool, PoolConfig config,
     this.host = host;
     this.socketTimeout = config.getSocketTimeoutMilliseconds();
     this.connectTimeout = config.getConnectTimeoutMilliseconds();
-    this.totalCount = 0;
+    this.totalCount = new AtomicInteger();
     this.log = new CustomLogger(name, host);
-    for (int i = 0; i < config.getMinSize(); i++) {
-      T object = objectFactory.create(host, socketTimeout, connectTimeout);
-      if (object != null) {
-        objectQueue.add(new Poolable<>(object, pool, host));
-        totalCount++;
+    this.takeSemaphore = new Semaphore(config.getMaxSize(), true);
+    try {
+      for (int i = 0; i < config.getMinSize(); i++) {
+          T object = objectFactory.create(host, socketTimeout, connectTimeout);
+          objectQueue.add(new Poolable<>(object, pool, host));
+          totalCount.incrementAndGet();
       }
     }
+    catch (Exception e) {
+      // skipping logging the exception as factories are already logging.
+    }
   }
 
   public void returnObject(Poolable<T> object)
   {
-    if (!objectFactory.validate(object.getObject())) {
-      log.debug(String.format("Invalid object...removing: %s ", object));
-      decreaseObject(object);
-      // Compensate for the removed object. Needed to prevent endless wait when in parallel a borrowObject is called
-      increaseObjects(1, false);
-      return;
-    }
+    try {
+      if (!objectFactory.validate(object.getObject())) {
+        log.debug(String.format("Invalid object...removing: %s ", object));
+        decreaseObject(object);
+        return;
+      }
 
-    log.debug(String.format("Returning object: %s to queue. Queue size: %d", object, objectQueue.size()));
-    if (!objectQueue.offer(object)) {
-      log.warn("Created more objects than configured. Created=" + totalCount + " QueueSize=" + objectQueue.size());
-      decreaseObject(object);
+      log.debug(String.format("Returning object: %s to queue. Queue size: %d", object, objectQueue.size()));
+      if (!objectQueue.offer(object)) {
+        log.warn("Created more objects than configured. Created=" + totalCount + " QueueSize=" + objectQueue.size());
```

Comment:
is it ok to release the semaphore incase more objects were created than `maxSize`

Reply:
No, this needs a change

Reply:
maybe I am missing something. Theoretically, this case should happen as we are controlling parallelism via counter-based semaphore. 
suppose, this happens, then why shouldn't we release the semaphore as the returning thread is done with the object? won't it give a chance to other waiting threads?

Reply:
if we are controlling parallelism strictly via counter-based semaphore, then this is an error state. We should error out if this happens instead.
I don't think it makes sense to operate a semaphore if the number of objects can be larger than permits (it can lead to wrong calculations).

Reply:
yes, true. throwing error in such a case.